### DISCUSSION TITLE
Suggestions on error of `6 | Connection refused`

### DISCUSSION THREAD

**Discussion by psychelzh** [r-lib/mirai#109] [Q&A] [answered]
Originally posted in https://github.com/wlandau/crew.cluster/discussions/38

Thanks for this great package, which is especially useful for crew package. The problem comes from using it with sge system, and these days I figured out that the problem might come from mirai, given these logged errors:

```
Error in dial(sock, url = url, autostart = asyncdial || NA, tls = tls,  : 
  6 | Connection refused
Calls: <Anonymous> ... do.call -> <Anonymous> -> dial_and_sync_socket -> dial
Execution halted
```

Thank you very much for any suggestions.

**Comment by shikokuchuo** [comment, marked as answer]
This error is simply that the daemons (workers) cannot connect back to the host. Two likely possibilities:

1. TLS is misconfigured - e.g. you are using your own certificate, but it is for a domain name which doesn't match that of the node you are connecting back to. Best way to rule out is by simply disabling TLS in crew.
2. There is a firewall around your login node that prevents inbound connections from the worker nodes.

Does the above help? For 2, I am guessing this may require your sysadmins to resolve.

**Reply by psychelzh** [reply]
Thank you very much for the suggestions. Here are what I got:

1. Tried to disable TLS by `crew_controller_sge(sge_log_output = "logs/sge", tls = crew::crew_tls("none"))`, but the error persists.
2. I do not know if there is such firewall, but I tried `ssh` back and forth between login node and worker nodes, which works fine. Does that mean there is no firewall?

I have previously succeeded in committing jobs in this way, only these days it is down. We have another login node, and things work great there.

**Reply by shikokuchuo** [reply]
If SSH works then it is even more likely a firewall issue - probably security was tightened up. SSH will be on the common port 22 or a specifically-configured port and those will typically be allowed (at least for internal IP ranges).

You can enquire if there is a whitelisted port range for your login node (to receive connections from other nodes), and then specify that to the 'port' argument of the crew controller. If not specified then a random high port range is used and those may well be blocked in environments such as yours.

**Reply by psychelzh** [reply]
Thanks, I know we can access through 9922 port. After specifying `port` argument, another error occurred:

```
Error in dial(sock, url = url, autostart = asyncdial || NA, tls = tls,  : 
  26 | Cryptographic error
Calls: <Anonymous> ... do.call -> <Anonymous> -> dial_and_sync_socket -> dial
Execution halted
```

Do you have any clue on this? Thank you very much.

**Reply by shikokuchuo** [reply]
Great, so this is a definite sign of the first type of issue. I assume that if you turn TLS off that it now works.

I'd like to refer to you the existing documentation on TLS: https://shikokuchuo.net/mirai/articles/mirai.html#distributed-computing-tls-secure-connections and I think there is also guidance at crew. If you'd like to discuss TLS configuration further, would you mind marking the answer here and opening a new discussion? That will help others with a similar problem find it. Thanks.

**Reply by psychelzh** [reply]
Thanks for the support! I will accept it as an answer and find out the problem with the TLS configuration.

**Reply by psychelzh** [reply]
Then I tried `crew_controller_sge(sge_log_output = "logs/sge", port = 9922, tls = crew::crew_tls("none"))`, which means that TLS is disabled. Then it throws a new error:

```
Error in dial(sock, url = url, autostart = asyncdial || NA, tls = tls,  : 
  13 | Protocol error
Calls: <Anonymous> ... do.call -> <Anonymous> -> dial_and_sync_socket -> dial
Execution halted
```

Does this mean I should set up a custom TLS configuration?

**Reply by shikokuchuo** [reply]
You seem to have run into another layer of issues! This one is not usual and will require some troubleshooting.

I suggest the following: if you are using `crew.cluster` (as it appears), then open a discussion there, and @ me so that I join the conversation. It might turn out to be an SGE thing, and you would be more likely to get an answer there. If you encounter this using just `mirai` then do open another discussion here – for example if you get this error when launching local daemons, or remote daemons over SSH tunnelling.

When you do, please include a minimal reprex if possible. Also details of your `sessionInfo()`, versions of `crew.cluster`, `crew`, `mirai` and `nanonext` and also the output of `nanonext::nng_version()`.

Also you mention one login node where everything works fine. If there are known differences between the two in terms of software environment or network configuration etc. then please also mention, as that may provide a clue.

**Reply by psychelzh** [reply]
Thanks! I will do troubleshooting first, and post whatever I found in a proper place.

**Reply by shikokuchuo** [reply]
Actually I have a theory as to what this is. Most likely the versions of `mirai` on  your login node and worker nodes have diverged. This is quite possible as `mirai` 0.13.2 was just released to CRAN a couple of days ago. If some have updated but some have not then you will run into this protocol error.

**Reply by psychelzh** [reply]
I have also thought about this possibility. However, given I am using `renv` to manage the versions of the packages, I found it is not an issue. Actually, I am still using `0.12.1` version of `mirai`, which is configured by the `renv` [lock file](https://github.com/psychelzh/cogstruct-dev/blob/main/renv.lock). Will this rule out that issue? Or will renv mess things up?

**Reply by shikokuchuo** [reply]
Is it possible for you to upgrade to `mirai` 0.13.2 everywhere? Version 0.12.1 suffers from an issue that is somewhat tangentially related and was fixed in 0.13.1. I don't think it will be possible to cleanly diagnose this current issue using that version.

Use of `renv` itself should be fine and I know there are `crew` users who use `renv`.

**Reply by psychelzh** [reply]
Great. That is an option. I will have a try on this.
